### PR TITLE
ARTEMIS-3679 Brokers shutdown after daylight saving fall back

### PR DESCRIPTION
**Pull request by franz1981**
https://issues.apache.org/jira/browse/ARTEMIS-3679

### CODE REVIEW

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLockTest.java [hunk 1]
```diff
@@ -116,6 +122,22 @@ public void dropLockTable() throws Exception {
...
    }
 
+   @Test
+   public void shouldStillHeldLockOnTimeZoneChange() {
+      Assume.assumeThat(PropertySQLProvider.Factory.identifyDialect(dbConf.getJdbcDriverClassName()), CoreMatchers.is(PropertySQLProvider.Factory.SQLDialect.DERBY));
+      TimeZone.setDefault(TimeZone.getTimeZone(ZoneId.of("-1")));
+      final long expirationMillis = Duration.ofMinutes(30).toMillis();
+      final LeaseLock lock = lock(expirationMillis);
+      Assert.assertTrue("Must acquire the lock!", lock.tryAcquire());
+      TimeZone.setDefault(TimeZone.getTimeZone(ZoneId.of("+3")));
```
Comment (clebertsuconic) [inline artemis-server/src/test/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLockTest.java:132]: this all look good to me, can't we merge this already?

Comment (franz1981) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLockTest.java:132]: Yep this is not ready to merge yet

Comment (franz1981) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLockTest.java]: @fvaleri I still have to decide how to handle testing this :)
But I've decided to use specific UTC get time query per supported DBs

File: artemis-jdbc-store/src/main/resources/journal-sql.properties [hunk 2]
```diff
@@ -75,14 +76,15 @@ create-file-table.mysql=CREATE TABLE %s(ID BIGINT NOT NULL AUTO_INCREMENT, FILEN
...
+current-timestamp.mysql=SELECT UTC_TIMESTAMP(6)
 # 4 GiB
 max-blob-size.mysql=4294967296
 table-names-case.mysql=lower
 
 # Oracle SQL statements
 create-file-table.oracle=CREATE TABLE %s(ID NUMBER(19) GENERATED BY DEFAULT ON NULL AS IDENTITY, FILENAME VARCHAR(255), EXTENSION VARCHAR(10), DATA BLOB, PRIMARY KEY(ID))
 create-journal-table.oracle=CREATE TABLE %s(id NUMBER(19),recordType NUMBER(5),compactCount NUMBER(5),txId NUMBER(19),userRecordType NUMBER(5),variableSize NUMBER(10),record BLOB,txDataSize NUMBER(10),txData BLOB,txCheckNoRecords NUMBER(10),seq NUMBER(19))
-
+current-timestamp.oracle=SELECT SYSTIMESTAMP AT TIME ZONE 'UTC' FROM DUAL
```
Comment (fvaleri) [inline artemis-jdbc-store/src/main/resources/journal-sql.properties:87]: Given the different behaviors and way to express these queries, I would say we should test them on all databases setting a system TZ different from UTC.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java [hunk 3]
```diff
@@ -114,14 +115,14 @@ private String readableLockStatus() {
          final boolean autoCommit = connection.getAutoCommit();
          connection.setAutoCommit(false);
          try (PreparedStatement preparedStatement = connection.prepareStatement(this.isLocked)) {
+            final long currentTimestamp = dbCurrentTimeMillis(connection);
```
Comment (fvaleri) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java:118]: I guess in dbCurrentTimeMillis() we should select the current UTC timestamp query based on the database type (e.g. `current-timestamp.mssql`) and fall back to `current-timestamp` with an appropriate warning if the database is unknown.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java [hunk 4]
```diff
@@ -147,14 +148,13 @@ private long dbCurrentTimeMillis(Connection connection) throws SQLException {
...
             resultSet.next();
             final long endTime = stripMilliseconds(System.currentTimeMillis());
-            final Timestamp currentTimestamp = resultSet.getTimestamp(1);
-            final long currentTime = currentTimestamp.getTime();
-            final long currentTimeMillis = stripMilliseconds(currentTime);
-            if (currentTimeMillis < startTime) {
-               LOGGER.warnf("[%s] %s query currentTimestamp = %s on database should happen AFTER %s on broker", lockName, holderId, currentTimestamp, new Timestamp(startTime));
+            final long currentTime = resultSet.getTimestamp(1, Calendar.getInstance(TimeZone.getTimeZone("UTC"))).getTime();
+            final long currentTimeNoMillis = stripMilliseconds(currentTime);
+            if (currentTimeNoMillis < startTime) {
```
Comment (fvaleri) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java:153]: Shouldn't this method be called something like verifyAndGetDatabaseCurrentTimestamp to better reflect the internal logic?

**Review by fvaleri — COMMENTED** [review]
Overall LGTM, but we need to carefully test the ad-hoc queries.

File: artemis-jdbc-store/src/main/resources/journal-sql.properties (hunk 2 above)
Comment (franz1981) [reply artemis-jdbc-store/src/main/resources/journal-sql.properties:90]: Agree, let's see what tests I can add

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java (hunk 4 above)
Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java:166]: Given that verification here won't imply anything but some output message probably `verify` can be omitted (unless I'll decide to perform a proper validation with thrown exception)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java (hunk 3 above)
Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java:120]: Yep, this is already how internally work the SQLProvider class donated by Jeff Mesnil from WildFly

Comment (fvaleri) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/jdbc/JdbcLeaseLock.java:120]: ok

**Pull request merged**